engine/generator.ts

```diff
@@ -0,0 +1,207 @@
+        let source = '';
+        if (block.members.length === 1) {
+            const member = block.members[0];
+            source += `layout(std140) uniform ${block.name} { ${this._genShaderVaryingNumObj(member)} };\n`;
+        } else {
+            source += `layout(std140) uniform ${block.name} {`;
+            for (const member of block.members) {
+                source += `    ${this._genShaderVaryingNumObj(member)}\n`;
+            }
+            source += `}\n`;
+        }
+        return source;
+    }
+    private _genShaderUBO4 (block: UniformBlock) : string {
+        let source = '';
+        if (block.members.length === 1) {
+            const member = block.members[0];
+            source += `layout(set = ${block.set}, binding = ${block.binding})`
+        + ` uniform ${block.name} { ${this._genShaderVaryingNumObj(member)} };\n`;
+        } else {
+            source += `layout(set = ${block.set}, binding = ${block.binding}) uniform ${block.name} {`;
+            for (const member of block.members) {
+                source += `    ${this._genShaderVaryingNumObj(member)}\n`;
+            }
+            source += `}\n`;
+        }
+        return source;
+    }
+
+    private _genShaderSampler1 (sampler: UniformSamplerTexture) : string {
+        return `uniform ${this._genShaderVaryingNumObj(sampler)}\n`;
+    }
+    private _genShaderSampler3 (sampler: UniformSamplerTexture) : string {
+        return `layout(std140) uniform ${this._genShaderVaryingNumObj(sampler)}\n`;
+    }
+    private _genShaderSampler4 (sampler: UniformSamplerTexture) : string {
+        return `layout(set = ${sampler.set}, binding = ${sampler.binding}) uniform ${this._genShaderVaryingNumObj(sampler)}\n`;
+    }
+
+    private _genShaderVarying1 (varying: IShaderVaryings, loc: number, source: IStandardShaderSource) : void {
+        source.vert += `varying out ${TypeInfos[varying.type].glslName} ${varying.name};\n`;
+        source.frag += `varying in ${TypeInfos[varying.type].glslName} ${varying.name};\n`;
+    }
+    private _genShaderVarying3 (varying: IShaderVaryings, loc: number, source: IStandardShaderSource) : void {
+        source.vert += `out ${TypeInfos[varying.type].glslName} ${varying.name};\n`;
+        source.frag += `in ${TypeInfos[varying.type].glslName} ${varying.name};\n`;
+    }
+    private _genShaderVarying4 (varying: IShaderVaryings, loc: number, source: IStandardShaderSource) : void {
+        source.vert += `layout (location = ${loc}) out ${TypeInfos[varying.type].glslName} ${varying.name};\n`;
+        source.frag += `layout (location = ${loc}) in ${TypeInfos[varying.type].glslName} ${varying.name};\n`;
+    }
+    private _genShaderBody1 (vert: string, frag: string, source: IStandardShaderSource): void {
+        source.vert += `${vert}
+            void main() {
+                gl_Position = vert();
+            }`;
+        source.frag += `${frag}
+            void main() {
+                gl_FragColor = u_color;
```

Comment:
This looks wrong. Should invoke `frag` and assign the result struct to the appropriate output.